### PR TITLE
Add Lift (NonEmpty instance)

### LINKED ISSUE #89: Backport Lift instance for NonEmpty
`template-haskell-2.15` (GHC 8.8) will [debut](http://git.haskell.org/ghc.git/commit/47875bd4d79ca633b589e63e320aa5a5c631d096) a `Lift` instance for `NonEmpty`. We should backport it in this library as well (behind a `template-haskell` Cabal flag).

### PR DESCRIPTION
**Pull request by phadej**
Resolves #89

### CODE REVIEW

File: src-ghc7/Data/List/NonEmpty.hs [hunk 1]
```diff
@@ -346,6 +350,24 @@ instance Foldable NonEmpty where
...
+        xs' <- TH.lift xs
+        return $ TH.InfixE (Just x') (TH.ConE neConsName) (Just xs')
+      where
+        neConsName = TH.mkNameG_d semigroupsPackageKey "Data.List.NonEmpty" ":|"
+
+semigroupsPackageKey         :: String
+#ifdef CURRENT_PACKAGE_KEY
+semigroupsPackageKey          = CURRENT_PACKAGE_KEY
+#else
+semigroupsPackageKey          = "lens-0.19" -- TODO: either remember to update or use Paths_
```
Comment (RyanGlScott) [inline src-ghc7/Data/List/NonEmpty.hs]: Yes, please do use `Paths_`—I don't know what this `lens` business is about :)

Comment (phadej) [reply src-ghc7/Data/List/NonEmpty.hs]: Ah, package name have to be hardcoded anyway :S

**Comment by RyanGlScott** [comment]
Other things to do:

* Update [`cabal.haskell-ci`](https://github.com/ekmett/semigroups/blob/c2f7ae6f42f0416bd5fae385f8a3c9232f355d94/cabal.haskell-ci) to include a `no-template-haskell` configuration set.
* Update the `CHANGELOG.md`.

**Comment by phadej** [comment]
on old GHC:

```
Language.Haskell.TH.Syntax: Can't be safely imported!
```

Just mark trustworthy?

**Comment by RyanGlScott** [comment]
> Just mark trustworthy?

I agree. You'll have to be careful, since recent versions of GHC (annoyingly) warn when a module marked `Trustworthy` could be inferred to be `Safe` is `-Wall` is enabled. This warning will likely fire on the `no-template-haskell` constraint set, so I'd just suppress that by adding these lines to `semigroups.cabal`:

```
  if impl(ghc >= 7.10)
    ghc-options: -fno-warn-trustworthy-safe
```

**Comment by phadej** [comment]
I didn't add `-fno-warn-trustworthy-safe` as "Recent GHC" doens't compile `Data.List.NonEmpty`

**Comment by RyanGlScott** [comment]
Very true! In that case, this looks good to me.

**Pull request merged**